src/Orleans/Async/TaskExtensions.cs

```diff
@@ -130,11 +131,13 @@ internal static T WaitForResultWithThrow<T>(this Task<T> task, TimeSpan timeout)
                 return;
             }
 
-            await Task.WhenAny(taskToComplete, Task.Delay(timeout));
+            var timeoutCancellationTokenSource = new CancellationTokenSource();
+            var completedTask = await Task.WhenAny(taskToComplete, Task.Delay(timeout, timeoutCancellationTokenSource.Token));
 
             // We got done before the timeout, or were able to complete before this code ran, return the result
-            if (taskToComplete.IsCompleted)
+            if (taskToComplete == completedTask)
```

Comment:
This will only cancel `Task.Delay()` upon a successful completion of `taskToComplete`. Shouldn't we also cancel it if `taskToComplete` failes, i.e. wrap the `await` in a `try`, catch any exceptions, cancel `Task.Delay()`, and rethrow? Or is it not worth it?